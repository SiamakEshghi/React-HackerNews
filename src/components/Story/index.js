import React, { PureComponent, Fragment } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import Spinner from 'components/Spinner';
import Ionicon from 'react-ionicons';
import * as actions from 'actions';
import "components/style.css";


class Story extends PureComponent {
    state = {
        loading: true
    }
    componentDidMount() {
        this.props.setTopStoryList( () => this.setState({loading: false}));
    }
    
    handleCommentList = (commentList) => {
        if(!commentList || commentList.length === 0) {
            alert("There is no comment for this story!")
        } else {
            this.props.setCommentList(commentList);
        }
    }
    renderStories = () => {
        const { topStoryList } = this.props;
        return (
            topStoryList.map((story) => {
                return (
                    <div key={story.id} className="row">
                        <Link 
                            className="btn" 
                            to={`/comment/${story.id}`} 
                            onClick={() => this.handleCommentList(story.kids)}
                        >
                            <Ionicon icon="md-text"  color="gray" />
                        </Link>
                        <h4>{story.title}</h4>
                    </div>
                );
            })
        );
    }
    render() {
        const { loading } = this.state;

        if (loading) {
            return (
                <div className="spinner">
                    <Spinner />
                </div>
            );
        } 
        return (
           <div className="container">
                <h2>Top Stories</h2>
                <div className="list">
                    {this.renderStories()}
                </div>
           </div>
        );
    }
}

const mapStateToProps = ({ reducer }) => {
    const { topStoryList } = reducer;

    return { topStoryList }
}
Story.propTypes = { topStoryList: PropTypes.array.isRequired };
export default connect(mapStateToProps, actions )(Story);