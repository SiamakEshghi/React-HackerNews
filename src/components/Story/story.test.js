import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import reduxThunk from 'redux-thunk';
import toJson from 'enzyme-to-json';
import Story from 'components/Story';

let wrapper, store;
const middlewares = [reduxThunk];
const mockstore = configureStore(middlewares);
const initialState = {reducer: {topStoryList: [{id: 1, title: "title1"}, {id: 2, title: "title2"}]}}

beforeEach(() => {
    store = mockstore(initialState);
    wrapper = mount(<Provider store={store}><Story/></Provider>);
});

afterEach(() => {
    wrapper.unmount();
});

describe('Story UI', () => {
    it('Should render correctly', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('check Props matches with initialState ', () => {
        expect(wrapper.find('Story').props().topStoryList).toEqual(initialState.reducer.topStoryList)
    });
});
