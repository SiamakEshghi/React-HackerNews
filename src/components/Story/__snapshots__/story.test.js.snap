// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Story UI Should render correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Story)>
    <Story
      setCommentList={[Function]}
      setTopStoryList={[Function]}
      topStoryList={
        Array [
          Object {
            "id": 1,
            "title": "title1",
          },
          Object {
            "id": 2,
            "title": "title2",
          },
        ]
      }
    >
      <div
        className="spinner"
      >
        <Component>
          <Loading
            color="#4A170C"
            delay={0}
            height={64}
            type="spokes"
            width={64}
          >
            <div
              dangerouslySetInnerHTML={
                Object {
                  "__html": "<svg id=\\"loading\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 32 32\\">
                  <path opacity=\\".1\\" d=\\"M14 0 H18 V8 H14 z\\" transform=\\"rotate(0 16 16)\\">
                    <animate attributeName=\\"opacity\\" from=\\"1\\" to=\\".1\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0\\"/>
                  </path>
                  <path opacity=\\".1\\" d=\\"M14 0 H18 V8 H14 z\\" transform=\\"rotate(45 16 16)\\">
                    <animate attributeName=\\"opacity\\" from=\\"1\\" to=\\".1\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0.125s\\"/>
                  </path>
                  <path opacity=\\".1\\" d=\\"M14 0 H18 V8 H14 z\\" transform=\\"rotate(90 16 16)\\">
                    <animate attributeName=\\"opacity\\" from=\\"1\\" to=\\".1\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0.25s\\"/>
                  </path>
                  <path opacity=\\".1\\" d=\\"M14 0 H18 V8 H14 z\\" transform=\\"rotate(135 16 16)\\">
                    <animate attributeName=\\"opacity\\" from=\\"1\\" to=\\".1\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0.375s\\"/>
                  </path>
                  <path opacity=\\".1\\" d=\\"M14 0 H18 V8 H14 z\\" transform=\\"rotate(180 16 16)\\">
                    <animate attributeName=\\"opacity\\" from=\\"1\\" to=\\".1\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0.5s\\"/>
                  </path>
                  <path opacity=\\".1\\" d=\\"M14 0 H18 V8 H14 z\\" transform=\\"rotate(225 16 16)\\">
                    <animate attributeName=\\"opacity\\" from=\\"1\\" to=\\".1\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0.675s\\"/>
                  </path>
                  <path opacity=\\".1\\" d=\\"M14 0 H18 V8 H14 z\\" transform=\\"rotate(270 16 16)\\">
                    <animate attributeName=\\"opacity\\" from=\\"1\\" to=\\".1\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0.75s\\"/>
                  </path>
                  <path opacity=\\".1\\" d=\\"M14 0 H18 V8 H14 z\\" transform=\\"rotate(315 16 16)\\">
                    <animate attributeName=\\"opacity\\" from=\\"1\\" to=\\".1\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0.875s\\"/>
                  </path>
                </svg>
                ",
                }
              }
              style={
                Object {
                  "fill": "#4A170C",
                  "height": 64,
                  "width": 64,
                }
              }
            />
          </Loading>
        </Component>
      </div>
    </Story>
  </Connect(Story)>
</Provider>
`;
